app/jobs/tweet_fetching_job.rb

```diff
@@ -0,0 +1,117 @@
+class TweetFetchingJob < ApplicationJob
+  queue_as :default
+
+  # These exceptions indicate that retry will be unsuccessful, so we should
+  # tell the parent DataSet that the user id has been fully processed.
+  rescue_from(Twitter::Error::Unauthorized,
+              Twitter::Error::BadRequest,
+              Twitter::Error::Forbidden,
+              Twitter::Error::NotFound,
+              Twitter::Error::NotAcceptable,
+              Twitter::Error::RequestEntityTooLarge,
+              Twitter::Error::UnprocessableEntity) do |exception|
+    do_completion_bookkeeping(false)
+  end
+
+  # This exception is retryable, but we should pause ALL jobs until after the
+  # rate limiting window resets.
+  rescue_from(Twitter::Error::TooManyRequests) do |exception|
+    # Stop execution of jobs; restart it after the rate limiting window has
+    # elapsed.
+    stop_cmd = "RAILS_ENV=#{Rails.env} #{Rails.root}/bin/delayed_job stop"
+    start_cmd = "RAILS_ENV=#{Rails.env} #{Rails.configuration.delayed_job_command} | at now + #{Rails.configuration.rate_limit_window} minutes"
```

Comment:
This is more complex and less portable than modifying the `run_at`s in the `delayed_jobs` table. Does it have advantages?

Reply:
Modifying `run_at` won't affect workers which have already reserved jobs (they can reserve multiple jobs at a time, and they check the value of `run_at` at the time of reservation).

That said if this is a graceful exit the workers are going to finish their queue before stopping, so I should check that behavior.

(Honestly the whole rate-limiting handling has been very challenging -- Twitter says "just check the endpoint periodically" but I couldn't think of a strategy that seemed sound for that, and even when I do that I need to stop everything once we hit the rate-limiting window anyway -- so if you have better ideas I am happy to hear them.)

Reply:
Spit balling: 
The fastest way to fulfill an arbitrary number of requests is to immediately use up all allowed requests in this 15 minute window, wait for window reset, and then immediately use up all allowed requests in the next 15 minute window, etc. To me, that suggests a single queue that maintains a running tally and that can be paused and resumed at any moment rather than maintaining a schedule. The architecture that falls out of that is a single, long running coordinator process that handles throttling the requests that get submitted to it. This can be a single worker running off a single queue. That worker doesn't actually run the requests. Instead, it creates request jobs that run out of a different queue on a different set of workers. If there are no requests left in this window's budget, the coordinator process sleeps until the window resets. Requests can still happen in parallel, but no request jobs get created if there is no budget for them.

I dunno. Does that sound reasonable? Things are way easier for me to reason about if there's a single choke point to handle throttling than trying to make decentralized workers coordinate.